l5kit/l5kit/kinematic/ackerman_perturbation.py

```diff
@@ -4,162 +4,108 @@
 
+    return trajectory
 
-def _compute_speeds_from_positions(trajectory: np.ndarray) -> np.ndarray:
-    xs = trajectory[:, 0]
-    ys = trajectory[:, 1]
-    speeds = np.zeros(xs.shape)
 
-    speeds[:-1] = np.sqrt((ys[1:] - ys[:-1]) ** 2 + (xs[1:] - xs[:-1]) ** 2)
-    speeds[-1] = speeds[-2]
-    return speeds
+def _compute_displacements(trajectory: np.ndarray) -> np.ndarray:
+    return np.linalg.norm(np.diff(trajectory[:, :2], axis=0), axis=1)
 
 
 class AckermanPerturbation(Perturbation):
-    def __init__(self, random_offset_generator: Callable, perturb_prob: float):
+    def __init__(
+        self,
+        random_offset_generator: Callable,
+        perturb_prob: float,
+        min_displacement: float,
+    ):
         """
         Apply Ackerman to get a feasible trajectory with probability perturb_prob.
 
         Args:
             random_offset_generator (RandomGenerator): a callable that yields 2 values
             perturb_prob (float): probability between 0 and 1 of applying the perturbation
+            min_displacement (float): minimum displacement required to apply lateral & yaw perturbation
         """
         self.perturb_prob = perturb_prob
+        self.min_displacement = min_displacement
         self.random_offset_generator = random_offset_generator
         if perturb_prob == 0:
             warnings.warn(
                 "Consider replacing this object with None if no perturbation is intended", RuntimeWarning, stacklevel=2
             )
 
     def perturb(
-            self, history_frames: np.ndarray, future_frames: np.ndarray, **kwargs: dict
+        self, history_frames: np.ndarray, future_frames: np.ndarray, **kwargs: dict
     ) -> Tuple[np.ndarray, np.ndarray]:
         if np.random.rand() >= self.perturb_prob:
             return history_frames.copy(), future_frames.copy()
 
-        lateral_offset_distance, yaw_offset_angle = self.random_offset_generator()
-
-        if np.abs(lateral_offset_distance) < NUMERICAL_THRESHOLD:
-            warnings.warn("ack not applied because of low lateral_distance", RuntimeWarning, stacklevel=2)
-            return history_frames.copy(), future_frames.copy()
+        (
+            lateral_offset_m,
+            longitudinal_offset_m,
+            yaw_offset_rad,
+        ) = self.random_offset_generator()
 
         num_history_frames = len(history_frames)
         num_future_frames = len(future_frames)
-        total_trajectory_length = num_history_frames + num_future_frames
-        if total_trajectory_length < 2:  # TODO is this an error?
-            #  we need at least 2 frames to compute speed and steering rate.
-            return history_frames.copy(), future_frames.copy()
 
-        new_trajectory_to_be_smoothed = _get_history_and_future_frames_as_joint_trajectory(
-            history_frames, future_frames
-        )
+        assert num_history_frames > 0, f"Number of history frames: {num_history_frames}"
```

Comment:
FYI this won't happen in the code, as current is always there even when you set `num_frames_history` to 0, but one assert has never killed anyone :)

Reply:
Yes, I know this will never goes to zero if we do not introduce BUG in agent sampling in the future. 
But I put assert here, because the perturbation code breaks when we do not have any history frames